task.go

```
@@ -2,69 +2,137 @@ package streams
+	}
+
+	t.setupTopology()
+
+	return nil
+}
 
-	t.stream = make(chan nodeMessage, 1000)
-	t.running.Set()
+func (t *streamTask) setupTopology() {
+	for _, node := range t.topology.Processors() {
+		pipe := NewProcessorPipe(node)
+		node.WithPipe(pipe)
 
-	ctx := NewProcessorPipe()
-	t.setupTopology(ctx)
+		ch := make(chan *Message, 1000)
+		t.procChs[node] = ch
+		t.procSigs[node] = t.runProcessor(node, ch)
+	}
 
-	t.consumeSources()
+	for source, node := range t.topology.Sources() {
+		pipe := NewProcessorPipe(node)
+		node.WithPipe(pipe)
 
-	for r := range t.stream {
-		ctx.SetNode(r.node)
-		if err := r.node.Process(r.msg); err != nil {
-			t.handleError(err)
-		}
+		t.runSource(source, node)
 	}
 }
 
-func (t *streamTask) setupTopology(ctx Pipe) {
-	for _, n := range t.topology.Processors() {
-		n.WithPipe(ctx)
-	}
+func (t *streamTask) runProcessor(node Node, ch chan *Message) chan bool {
+	done := make(chan bool, 1)
+
+	go func() {
+		for msg := range ch {
+			start := time.Now()
+
+			nodeMsgs, err := node.Process(msg)
+			if err != nil {
+				t.handleError(err)
+			}
+
+			stats.Timing(msg.Ctx, "node.latency", time.Since(start), 1.0, "name", node.Name())
+			stats.Inc(msg.Ctx, "node.throughput", 1, 1.0, "name", node.Name())
+			stats.Gauge(msg.Ctx, "node.back-pressure", pressure(ch), 0.1, "name", node.Name())
+
+			for _, nodeMsg := range nodeMsgs {
+				ch := t.procChs[nodeMsg.Node]
+				ch <- nodeMsg.Msg
+			}
+		}
+
+		done <- true
+	}()
+
+	return done
+}
+
+func (t *streamTask) runSource(source Source, node Node) {
+	go func() {
+		t.sourceWg.Add(1)
+		defer t.sourceWg.Done()
+
+		for t.running.IsSet() {
+			start := time.Now()
+
+			msg, err := source.Consume()
+			if err != nil {
+				t.handleError(err)
+			}
+
+			if msg.Empty() {
+				continue
+			}
+
+			stats.Timing(msg.Ctx, "node.latency", time.Since(start), 1.0, "name", node.Name())
+			stats.Inc(msg.Ctx, "node.throughput", 1, 1.0, "name", node.Name())
+
+			nodeMsgs, _ := node.Process(msg)
```

Why are you ignoring the error here?